### PR TITLE
Implement the File() function to read file as a String

### LINKED ISSUE #18851: A function to read file as a String.
```
SELECT file(path)
```

Restrictions on file path must apply (server can only read files from the `user_files` directory)

**Use case**

```
clickhouse-client --query "INSERT INTO table SELECT file('a.txt'), file('b.txt')"
```

```
clickhouse-local --query "SELECT file('a.txt'), file('b.txt')" | clickhouse-client --query "INSERT INTO table FORMAT TSV"
```

In this example, the whole content of the `a.txt` and `b.txt` file is inserted as a single String values.

**Notes**

We get used to the fact that storing large BLOBs in ClickHouse is discouraged.
But after enabling adaptive index granularity it becomes more ok.

### PR DESCRIPTION
**Pull request by keen-wolf**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added file() function to read file as a String.  This close [#issue:18851](https://github.com/ClickHouse/ClickHouse/issues/18851). 

Detailed description / Documentation draft:
Firstly get the file's absoluted path, then check if the access is allowed(in the SERVER mode only files within "user_files_path" can be read), at last, read file content directly into the ColumnString and return.

### CODE REVIEW

**Comment by keen-wolf** [comment]
Sorry for having deleted the “CMakeLists.txt” from git by misoperation. This request code does nothing with it

**Comment by keen-wolf** [comment]
Sorry for having deleted the “CMakeLists.txt” from git by misoperation. This request code does nothing with it

**Review by qoega — COMMENTED** [review]
How do we control that this function will not allow to read data from other databases and private configs? It is very powerful function as it runs with the same user as clickhouse-server.

**Comment by alex-zaitsev** [comment]
> How do we control that this function will not allow to read data from other databases and private configs? It is very powerful function as it runs with the same user as clickhouse-server.

https://clickhouse.tech/docs/en/operations/server-configuration-parameters/settings/#server_configuration_parameters-user_files_path ?

**Comment by keen-wolf** [comment]
> How do we control that this function will not allow to read data from other databases and private configs? It is very powerful function as it runs with the same user as clickhouse-server.

yes，as @alex-zaitsev has pointed, now the file function in this PR does the same path checking as the file() in Table Functions. The path is restricted to only the scope of "user_files_path" being configured in the server config file, elsewhere is not allowed.

But in this function’s test, I also found as the user of clickhouse-server do have large powers which can access almost everything especially when you run it as "root", this will be dangerous. So when handle these low-level functions such as file(), which can directly interact with OS or syscalls, it should be treated very careful.

Also I found other file access behavior in clickhouse as shown in attached screenshot. In this simple test the “outfile “ command seems to having the authority to write file anywhere without limitation, but when read with the file function, it was refused. I don’t know if this is the primitive design or is some kind of leakage.

image

**Comment by alexey-milovidov** [comment]
> Also I found other file access behavior in clickhouse as shown in attached screenshot. In this simple test the “outfile “ command seems to having the authority to write file anywhere without limitation, but when read with the file function, it was refused. I don’t know if this is the primitive design or is some kind of leakage.

INTO OUTFILE is processed by client, not the server.
It has the same rights as the user running `clickhouse-client`.

**Comment by alexey-milovidov** [comment]
Also there is the case:
- if the query is processed by `clickhouse-local`, restriction should not apply.

File: src/Functions/FunctionFile.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+        DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+        {
+            if (!isString(arguments[0].type))
+                throw Exception(getName() + " is only implemented for types String", ErrorCodes::NOT_IMPLEMENTED);
+            return std::make_shared<DataTypeString>();
+        }
+
+        bool useDefaultImplementationForConstants() const override { return true; }
+        ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {1}; }
```
Comment (alexey-milovidov) [inline src/Functions/FunctionFile.cpp:48]: I see that non-constant argument is supported (as expected). So, this is line is unneeded.

File: src/Functions/FunctionFile.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                throw Exception("Bad Function arguments for file() " + std::string(filename), ErrorCodes::ILLEGAL_COLUMN);
+            }
+        }
+
+    private:
+        void checkReadIsAllowed(const std::string & user_files_path, const std::string & file_path) const
+        {
+            // If run in Local mode, no need for path checking.
+            if (context.getApplicationType() != Context::ApplicationType::LOCAL)
+                if (file_path.find(user_files_path) != 0)
```
Comment (alexey-milovidov) [inline src/Functions/FunctionFile.cpp:108]: Unsafe.

Comment (alexey-milovidov) [reply src/Functions/FunctionFile.cpp:108]: Path traversal vulnerability.

Comment (keen-wolf) [reply src/Functions/FunctionFile.cpp:108]: ok, I'll reinforce it

File: src/Functions/FunctionFile.cpp (hunk 1 above)
Comment (keen-wolf) [reply src/Functions/FunctionFile.cpp:48]: Remove it soon.

**Comment by keen-wolf** [comment]
> > Also I found other file access behavior in clickhouse as shown in attached screenshot. In this simple test the “outfile “ command seems to having the authority to write file anywhere without limitation, but when read with the file function, it was refused. I don’t know if this is the primitive design or is some kind of leakage.
> 
> INTO OUTFILE is processed by client, not the server.
> It has the same rights as the user running `clickhouse-client`.

got it！

**Comment by alex-zaitsev** [comment]
@keen-wolf , @alexey-milovidov , do we really need another function if there is RawBLOB Format?

**Comment by keen-wolf** [comment]
> @keen-wolf , @alexey-milovidov , do we really need another function if there is RawBLOB Format?

As I can see, the file function can get more flexibility when use it as a single column which can be placed freely in the SQL, while the RawBLOB can only be used with local pipes at the ClickHouse client. This difference I think comes from the mechanism between Format and Function. This also authorize the file() to read files from the server node while RawBLOB can’t. 
But I can’t give a rigid scenario in which must use this feature. Maybe we can consider to use it together with the table file function which can write data to the server node, this may produce a closed loop to form some interesting abilities or tricky skills.

**Comment by alexey-milovidov** [comment]
@nikitamikhaylov Multiple bugs were found: https://github.com/ClickHouse/ClickHouse/pull/32640

**Pull request merged**